apps/prairielearn/elements/pl-file-upload/pl-file-upload.py

```diff
@@ -37,48 +110,89 @@ def add_format_error(data: pl.QuestionData, error_string: str) -> None:
 
 def prepare(element_html: str, data: pl.QuestionData) -> None:
     element = lxml.html.fragment_fromstring(element_html)
-    required_attribs = ["file-names"]
-    optional_attribs = []
+
+    # Either "file-names" or "optional-file-names" is required, which is checked separately
+    required_attribs = []
+    optional_attribs = ["file-names", "optional-file-names", "allow-blank"]
     pl.check_attribs(element, required_attribs, optional_attribs)
+    if not pl.has_attrib(element, "file-names") and not pl.has_attrib(
+        element, "optional-file-names"
+    ):
+        raise Exception(
+            "One of required attributes file-names or optional-file-names missing"
+        )
+
+    if pl.has_attrib(element, "file-names") and pl.has_attrib(element, "allow-blank"):
+        raise Exception(
+            "The attribute allow-blank cannot be used when (mandatory) file-names are specified"
+        )
 
     if "_required_file_names" not in data["params"]:
         data["params"]["_required_file_names"] = []
-    file_names = get_file_names_as_array(pl.get_string_attrib(element, "file-names"))
-    data["params"]["_required_file_names"].extend(file_names)
 
+    required_file_names = get_file_names_as_array(
+        pl.get_string_attrib(element, "file-names", "")
+    )
+    data["params"]["_required_file_names"].extend(required_file_names)
 
-def render(element_html: str, data: pl.QuestionData) -> str:
-    if data["panel"] != "question":
-        return ""
 
+def render(element_html: str, data: pl.QuestionData) -> str:
     element = lxml.html.fragment_fromstring(element_html)
     uuid = pl.get_uuid()
+    parse_error = data["format_errors"].get("_files", None)
```

Comment:
```suggestion
    parse_error = data["format_errors"].get("_files")
```
`None` is used as the default value if not specified.